Initial implementation of Redis Streams.

ttingen opened the pull request:
The Streams data type is available in release 5.0 RC1 and above.

- Implemented Sync & Async methods for all Stream related commands (minus the blocking options) as of 5.0 RC1.
- Added tests for the synchronous versions of the Streams API but the testing is a work in progress. Need to refactor for reuse within the streams tests and write a more thorough suite of tests.
- Added a NameValueEntry struct which mimicks HashEntry. Using HashEntry for the name/value pairs of stream entries seemed wrong. Perhaps refactor the usage of HashEntry to the more generic NameValueEntry and deprecate HashEntry?

mgravell commented:
Oh, awesome. I can see I have quite some reading ahead, but I guess that's
better than reading + writing :)

Can I confirm that you are able and willing to freely contribute this PR
under the project's license? Essentially the "no I won't sue you and claim
I own that part of the code later" legals check?

On Sun, 17 Jun 2018, 05:32 Todd Tingen, <notifications@github.com> wrote:

> The Streams data type is available in release 5.0 RC1 and above.
>
>    - Implemented Sync & Async methods for all Stream related commands
>    (minus the blocking options) as of 5.0 RC1:
>    XACK,
>    XADD,
>    XCLAIM,
>    XDEL,
>    XGROUP,
>    XINFO,
>    XLEN,
>    XPENDING,
>    XRANGE,
>    XREAD,
>    XREADGROUP,
>    XREVRANGE,
>    XTRIM
>    - Added tests for the synchronous versions of the Streams API but the
>    testing is a work in progress. Need to refactor for reuse within the
>    streams tests and write a thorough suite of tests.
>    - Added a NameValueEntry struct which mimicks HashEntry. Using
>    HashEntry for the name/value pairs of stream entries seemed wrong. Perhaps
>    refactor the usage of HashEntry to the more generic NameValueEntry and
>    deprecate HashEntry?
>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/StackExchange/StackExchange.Redis/pull/860
> Commit Summary
>
>    - Initial implementation of Redis Streams.
>
> File Changes
>
>    - *M* StackExchange.Redis.Tests/DatabaseWrapperTests.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/860/files#diff-0>
>    (197)
>    - *M* StackExchange.Redis.Tests/StackExchange.Redis.Tests.csproj
>    <https://github.com/StackExchange/StackExchange.Redis/pull/860/files#diff-1>
>    (4)
>    - *A* StackExchange.Redis.Tests/Streams.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/860/files#diff-2>
>    (832)
>    - *M* StackExchange.Redis.Tests/WrapperBaseTests.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/860/files#diff-3>
>    (197)
>    - *M* StackExchange.Redis/StackExchange/Redis/Enums/RedisCommand.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/860/files#diff-4>
>    (14)
>    - *M* StackExchange.Redis/StackExchange/Redis/ExtensionMethods.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/860/files#diff-5>
>    (12)
>    - *M* StackExchange.Redis/StackExchange/Redis/Interfaces/IDatabase.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/860/files#diff-6>
>    (314)
>    - *M*
>    StackExchange.Redis/StackExchange/Redis/Interfaces/IDatabaseAsync.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/860/files#diff-7>
>    (312)
>    - *M*
>    StackExchange.Redis/StackExchange/Redis/KeyspaceIsolation/DatabaseWrapper.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/860/files#diff-8>
>    (135)
>    - *M*
>    StackExchange.Redis/StackExchange/Redis/KeyspaceIsolation/WrapperBase.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/860/files#diff-9>
>    (135)
>    - *A* StackExchange.Redis/StackExchange/Redis/NameValueEntry.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/860/files#diff-10>
>    (85)
>    - *M* StackExchange.Redis/StackExchange/Redis/RedisDatabase.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/860/files#diff-11>
>    (771)
>    - *M* StackExchange.Redis/StackExchange/Redis/RedisFeatures.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/860/files#diff-12>
>    (8)
>    - *A* StackExchange.Redis/StackExchange/Redis/RedisStream.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/860/files#diff-13>
>    (32)
>    - *A* StackExchange.Redis/StackExchange/Redis/RedisStreamEntry.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/860/files#diff-14>
>    (32)
>    - *M* StackExchange.Redis/StackExchange/Redis/ResultProcessor.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/860/files#diff-15>
>    (478)
>    - *A* StackExchange.Redis/StackExchange/Redis/StreamConstants.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/860/files#diff-16>
>    (59)
>    - *A* StackExchange.Redis/StackExchange/Redis/StreamConsumer.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/860/files#diff-17>
>    (19)
>    - *A* StackExchange.Redis/StackExchange/Redis/StreamConsumerInfo.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/860/files#diff-18>
>    (32)
>    - *A* StackExchange.Redis/StackExchange/Redis/StreamGroupInfo.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/860/files#diff-19>
>    (32)
>    - *A* StackExchange.Redis/StackExchange/Redis/StreamInfo.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/860/files#diff-20>
>    (54)
>    - *A* StackExchange.Redis/StackExchange/Redis/StreamPendingInfo.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/860/files#diff-21>
>    (41)
>    - *A*
>    StackExchange.Redis/StackExchange/Redis/StreamPendingMessageInfo.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/860/files#diff-22>
>    (43)
>
> Patch Links:
>
>    - https://github.com/StackExchange/StackExchange.Redis/pull/860.patch
>    - https://github.com/StackExchange/StackExchange.Redis/pull/860.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/StackExchange/StackExchange.Redis/pull/860>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AABDsGijegWKWZt5Bg4nKODgqa3MXHzCks5t9dv6gaJpZM4UqtE5>
> .
>

NickCraver commented:
Lots of reading to do indeed! I just went through a quick cursory scan of this and I wanted to say: this is **fantastic**.

It's the largest first time contribution I've seen to any project and also appears to be very well polished. **Thank you**. Marc and I were already talking about adding this and the other APIs we haven't caught up with (including overloads) in v2 that we're working on now. This is perfect timing too. Marc's busy on the pipelines branch at the moment but we'll give this a shakedown and get it in ASAP. The test suite helps immensely here. Again: very well done. I'm impressed.

ttingen commented:
@mgravell Yes, I'm willling to contribute this PR under the project's license.

Thanks for the positive feedback, streams is a feature that I look forward to using. 

I already see some things that my hasty review last night at midnight (EDT) missed. Copypasta got me on some of the unit test comments and the XML documentation on the NameValueEntry struct still references HashEntry, those things need to be cleaned up.

I also didn't like the naming of some of the API methods and parameters (e.g., StreamClaimMessagesReturningIds, and some others).  The names are sufficiently descriptive but I think they could be improved. Naming is hard...

NickCraver commented:
Need to go through code next, but have 5.0-rc3 up and running and test suite against it here via WSL:
![image](https://user-images.githubusercontent.com/454813/41881375-366be744-78b1-11e8-8d6c-cef4fe10ff95.png)
![image](https://user-images.githubusercontent.com/454813/41881376-37f48c10-78b1-11e8-8c21-d741e467444f.png)

I'll start on code review right now, but will take a bit since you've done so much work here :)

NickCraver commented on StackExchange.Redis/StackExchange/Redis/RedisStream.cs:
```diff
@@ -0,0 +1,32 @@
+        /// <summary>
+        /// An arry of entries contained within the stream.
+        /// </summary>
+        public RedisStreamEntry[] Entries => entries;
```
Can simplify this and other structs down to `{ get; }`-only properties, e.g. `public RedisStreamEntry[] Entries { get; }` here.

NickCraver commented on StackExchange.Redis/StackExchange/Redis/ResultProcessor.cs:
```diff
@@ -1301,6 +1325,461 @@ protected override bool SetResultCore(PhysicalConnection connection, Message mes
+                    //             2) "value_2"
+
+                    result = result.GetItems()[0]  // Initial result array, then the first stream.
+                                   .GetItems()[1]; // Skip the stream name and return the array of stream entries.
```
Should use a local for `result.GetItems()` here and access it twice.

NickCraver commented on StackExchange.Redis/StackExchange/Redis/ResultProcessor.cs:
```diff
@@ -1301,6 +1325,461 @@ protected override bool SetResultCore(PhysicalConnection connection, Message mes
+                if (result.IsNull)
+                {
+                    // Server returns 'nil' if no entries are returned for the given stream.
+                    SetResult(message, new RedisStreamEntry[0]);
```
Use `Array.Empty<RedisStreamEntry>()` instead here

NickCraver commented on StackExchange.Redis/StackExchange/Redis/ResultProcessor.cs:
```diff
@@ -1301,6 +1325,461 @@ protected override bool SetResultCore(PhysicalConnection connection, Message mes
+                if (result.IsNull)
+                {
+                    // Nothing returned for any of the requested streams. The server returns 'nil'.
+                    SetResult(message, new RedisStream[0]);
```
`Array.Empty<RedisStream>()`

NickCraver commented on StackExchange.Redis/StackExchange/Redis/ResultProcessor.cs:
```diff
@@ -1301,6 +1325,461 @@ protected override bool SetResultCore(PhysicalConnection connection, Message mes
+                for (var i = 0; i < arr.Length; i++)
+                {
+                    parsedItems[i] = ParseItem(arr[i]);
+                }
```
Can simplify this and other similar loops using `Array.ConverAll`, e.g.
```c#
var parsedItems = Array.ConvertAll(arr, ParseItem);
```

mgravell replied:
Nit for @NickCraver: prefer:

```
var parsedItems = Array.ConvertAll(arr, item => ParseItem(item));
```

Unless I'm mistaken, they're still subtly different. Happy to find out they fixed the compiler, of course.

NickCraver commented on StackExchange.Redis/StackExchange/Redis/ResultProcessor.cs:
```diff
@@ -1301,6 +1325,461 @@ protected override bool SetResultCore(PhysicalConnection connection, Message mes
+                {
+                    arr[9],
+                    arr[11]
+                }));
```
This one is a bit round-about to get the parsed result - we should probably break up the parsing so it's accessible/this is more direct without the intermediate creations.

ttingen replied:
Something like this?

![image](https://user-images.githubusercontent.com/1170315/41945113-6a899418-7979-11e8-9dd3-26c112027a98.png)

ttingen replied:
Something like this? 

![image](https://user-images.githubusercontent.com/1170315/41945194-cd89f0ee-7979-11e8-98a9-527154cc5072.png)

ttingen replied:
This looks cleaner than the previous version. Another option would be to specify the constructor parameters by name instead of with the comment. I like to have an explicit identifier (named param or comment) of what it is I'm passing if it's not readily apparent from the value being passed (just a personal preference).

NickCraver commented on StackExchange.Redis/StackExchange/Redis/ResultProcessor.cs:
```diff
@@ -1301,6 +1325,461 @@ protected override bool SetResultCore(PhysicalConnection connection, Message mes
+
+                if (count == 0)
+                {
+                    pairs = new NameValueEntry[0];
```
`Array.Empty<NameValueEntry>()`

Also: would just `return` directly here and in the else, keep it simple!

Note: it's okay to drop `net45` from the `.csproj` to get this to compile...we're already on `netstandard2.0`-only in the `pipelines` branch already, so no loss :)

NickCraver commented on StackExchange.Redis/StackExchange/Redis/StreamConstants.cs:
```diff
@@ -0,0 +1,59 @@
+        /// <summary>
+        /// The "+" value used in the XRANGE, XREAD, and XREADGROUP commands. Indicates the maximum message ID from the stream.
+        /// </summary>
+        public static readonly RedisValue StreamMaxValue = "+";
```
Naming: suggest `ReadMinValue`/`ReadMaxValue` (rather than repeating `Stream`), thoughts?

NickCraver commented on StackExchange.Redis/StackExchange/Redis/StreamConsumer.cs:
```diff
@@ -0,0 +1,19 @@
+    /// <summary>
+    /// Describes a consumer off a Redis Stream.
+    /// </summary>
+    public class StreamConsumer
```
Would make this a `struct` (with constructor)

NickCraver commented on StackExchange.Redis/StackExchange/Redis/RedisStreamEntry.cs:
```diff
@@ -0,0 +1,42 @@
+    public struct RedisStreamEntry
+    {
+        internal readonly RedisValue id;
+        internal readonly NameValueEntry[] values;
```
Can use `{ get; }`-only properties here

NickCraver commented on StackExchange.Redis/StackExchange/Redis/RedisDatabase.cs:
```diff
@@ -2164,6 +2805,171 @@ private Message GetSortedSetRemoveRangeByScoreMessage(RedisKey key, double start
+            // Example: > XREADGROUP GROUP mygroup Alice COUNT 1 STREAMS mystream >
+            if (count.HasValue && count <= 0)
+            {
+                throw new ArgumentOutOfRangeException("count must be greater than 0.");
```
`throw new ArgumentOutOfRangeException(nameof(count), "count must be greater than 0.");`
(current code is calling the `string paramName` overload)

NickCraver commented on StackExchange.Redis/StackExchange/Redis/RedisDatabase.cs:
```diff
@@ -2164,6 +2805,171 @@ private Message GetSortedSetRemoveRangeByScoreMessage(RedisKey key, double start
+
+            if (count <= 0)
+            {
+                throw new ArgumentOutOfRangeException("count must be greater than 0.");
```
`throw new ArgumentOutOfRangeException(nameof(count), "count must be greater than 0.");`

NickCraver commented on StackExchange.Redis/StackExchange/Redis/RedisDatabase.cs:
```diff
@@ -2164,6 +2805,171 @@ private Message GetSortedSetRemoveRangeByScoreMessage(RedisKey key, double start
+            // See https://redis.io/commands/xadd.
+
+            if (streamFields == null) throw new ArgumentNullException(nameof(streamFields));
+            if (streamFields.Length == 0) throw new ArgumentOutOfRangeException("streamFields must contain at least one field.");
```
`throw new ArgumentOutOfRangeException(nameof(streamFields), "streamFields must contain at least one field.");`

NickCraver commented on StackExchange.Redis/StackExchange/Redis/RedisDatabase.cs:
```diff
@@ -2164,6 +2805,171 @@ private Message GetSortedSetRemoveRangeByScoreMessage(RedisKey key, double start
+
+            if (maxLength.HasValue && maxLength <= 0)
+            {
+                throw new ArgumentOutOfRangeException("maxLength must be greater than 0.");
```
`throw new ArgumentOutOfRangeException(nameof(maxLength), "maxLength must be greater than 0.");`

NickCraver commented on StackExchange.Redis/StackExchange/Redis/RedisDatabase.cs:
```diff
@@ -1579,6 +1579,605 @@ public IEnumerable<SortedSetEntry> SortedSetScan(RedisKey key, RedisValue patter
             return ExecuteAsync(msg, ResultProcessor.NullableDouble);
         }
 
+        public long StreamAcknowledge(RedisKey key, RedisValue groupName, string messageId, CommandFlags flags = CommandFlags.None)
```
I think `messageId` (here and elsewhere - returns included) should be a `RedisValue` going in (for consistency as well), since we're converting underneath on every path used. Thoughts?

NickCraver commented on StackExchange.Redis/StackExchange/Redis/RedisDatabase.cs:
```diff
@@ -1579,6 +1579,605 @@ public IEnumerable<SortedSetEntry> SortedSetScan(RedisKey key, RedisValue patter
+        {
+            var msg = GetStreamPendingMessagesMessage(key, groupName, minId, maxId, count, consumerName, flags);
+            return ExecuteAsync(msg, ResultProcessor.StreamPendingMessages);
+        }
```
nit: can we keep sync/async pattern on methods here? last 4 aren't paired up

mgravell replied:
I thought there used to be a test that checked this with reflection... If not, I can add

ttingen replied:
I'm not sure I understand this one, there are sync/async methods for the StreamPendingMessageInfoGet operation. Or is the naming too similar to StreamPendingInfoGet?

StreamPendingInfoGet returns a summary of pending information for the stream, whereas StreamPendingMessageInfoGet returns the information for each pending message.

![image](https://user-images.githubusercontent.com/1170315/41947876-2f474832-7988-11e8-80e9-71277c46f044.png)

NickCraver replied:
In this case there are 2 issues:
- Ordering (should be method, async equivalent, method, async equivalent) for consistency. 
- I don't think we need 2 sets of methods here - if consumer name is null, we don't return details (just default it to `RedisValue.Null`)
...that gets the same functionality while simplifying the API. We should document this in the usage though.

NickCraver commented on StackExchange.Redis/StackExchange/Redis/RedisDatabase.cs:
```diff
@@ -1579,6 +1579,605 @@ public IEnumerable<SortedSetEntry> SortedSetScan(RedisKey key, RedisValue patter
+            return ExecuteAsync(msg, ResultProcessor.StreamPendingMessages);
+        }
+
+        public RedisStreamEntry[] StreamRange(RedisKey key, RedisValue minId, RedisValue maxId, CommandFlags flags = CommandFlags.None)
```
Same for below - instead of 2 methods - how about we make `count` an `int?` on these? That'd take us from 4 (sync+async) methods down to 2.

NickCraver commented on StackExchange.Redis/StackExchange/Redis/RedisDatabase.cs:
```diff
@@ -1579,6 +1579,605 @@ public IEnumerable<SortedSetEntry> SortedSetScan(RedisKey key, RedisValue patter
+            return ExecuteSync(msg, ResultProcessor.SingleStream);
+        }
+
+        public RedisStreamEntry[] StreamRangeReverse(RedisKey key, RedisValue maxId, RedisValue minId, CommandFlags flags = CommandFlags.None)
```
Same as above - by naming `count` nullable, we can collapse overloads here and simplify

Overall: for API consistency instead of `StreamRangeReverse`, look at how `SortedSetRangeByRank` and others work - add a `Order order = Order.Ascending` argument in the same way - this again halves our overloads needed :)

NickCraver commented on StackExchange.Redis/StackExchange/Redis/RedisDatabase.cs:
```diff
@@ -1579,6 +1579,605 @@ public IEnumerable<SortedSetEntry> SortedSetScan(RedisKey key, RedisValue patter
+            return ExecuteAsync(msg, ResultProcessor.SingleStreamWithNameSkip);
+        }
+
+        public Task<RedisStreamEntry[]> StreamReadAsync(RedisKey key, RedisValue afterId, int count, CommandFlags flags = CommandFlags.None)
```
Same as elsewhere: `int? count` and collapse overloads

NickCraver commented on StackExchange.Redis/StackExchange/Redis/RedisDatabase.cs:
```diff
@@ -1579,6 +1579,605 @@ public IEnumerable<SortedSetEntry> SortedSetScan(RedisKey key, RedisValue patter
+            return ExecuteAsync(msg, ResultProcessor.MultiStream);
+        }
+
+        public Task<RedisStream[]> StreamReadAsync(IList<KeyValuePair<RedisKey, RedisValue>> streamWithAfterIdList, int countPerStream, CommandFlags flags = CommandFlags.None)
```
Collapse: `int? count` with above.

ttingen commented:
Lots to update. I will start on these changes tomorrow evening.

mgravell commented on StackExchange.Redis/StackExchange/Redis/RedisDatabase.cs:
```diff
@@ -2164,6 +2805,171 @@ private Message GetSortedSetRemoveRangeByScoreMessage(RedisKey key, double start
                     GetRange(start, exclude, true), GetRange(stop, exclude, false));
         }
 
+        private Message GetStreamAcknowledgeMessage(RedisKey key, RedisValue groupName, CommandFlags flags, params string[] messageIds)
```
Passing query: is messageId a strict string? Or should it be RedisValue?

ttingen commented on StackExchange.Redis/StackExchange/Redis/Interfaces/IDatabase.cs:
```diff
@@ -1389,6 +1389,318 @@ public interface IDatabase : IRedis, IDatabaseAsync
+        /// <param name="useApproximateMaxLength">If true, use the "~" argument to keep the stream length at or slightly above the maxLength.</param>
+        /// <returns>The ID of the newly created message.</returns>
+        /// <remarks>https://redis.io/commands/xadd</remarks>
+        string StreamAdd(RedisKey key, NameValueEntry[] streamFields, int? maxLength = null, bool useApproximateMaxLength = false, CommandFlags flags = CommandFlags.None);
```
I like streamPairs, will go with that.

ttingen commented:
Just commited many of the smaller items from the change request. I will try to knock out the remainder of them tomorrow.

I still need to do the following:
- Reduce the number of overloads.
- Remove net45 target to allow Array.Empty usage and then make those changes.
- Convert messageId from string to RedisValue.

NickCraver commented:
Note: I pushed a launch-all-servers bash script in https://github.com/StackExchange/StackExchange.Redis/commit/18c607d463a259271f34dfc3b3d21a334943c742 that I'm using to fire up all servers in WSL for testing. This should be useful for anyone testing this PR. @ttingen If you can yank in master on the next pass, that'll make life easier :)

Appreciate all the work here - looking good.

ttingen commented:
I believe I completed all of the requested changes (minus the Array.Empty changes). I added a few more tests and I will continue that effort.

I didn't collapse any of the StreamAdd overloads but I would like to remove half of them. If the streamEntryId parameter were a string, we could default it to "*". But I understand the reasoning to want to standardize on using the RedisValue type for those values.

NickCraver commented on StackExchange.Redis/StackExchange/Redis/Interfaces/IDatabase.cs:
```diff
@@ -1389,6 +1389,250 @@ public interface IDatabase : IRedis, IDatabaseAsync
+        /// <param name="useApproximateMaxLength">If true, the "~" argument is used to allow the stream to exceed max length by a small number. This improves performance when removing messages.</param>
+        /// <returns>The ID of the newly created message.</returns>
+        /// <remarks>https://redis.io/commands/xadd</remarks>
+        RedisValue StreamAdd(RedisKey key, RedisValue streamEntryId, RedisValue streamField, RedisValue streamValue, int? maxLength = null, bool useApproximateMaxLength = false, CommandFlags flags = CommandFlags.None);
```
How about we move `streamEntryId` as the first optional arg (on both sync and async)? That'll eliminate down 4 overloads in a hurry (2 on single fields, 2 on pairs overloads).

ttingen replied:
Change the type to a string since we can’t default to RedisValue.Null?

NickCraver commented on StackExchange.Redis/StackExchange/Redis/Interfaces/IDatabase.cs:
```diff
@@ -1389,6 +1389,250 @@ public interface IDatabase : IRedis, IDatabaseAsync
+        /// <param name="flags">The flags to use for this operation.</param>
+        /// <returns>A <see cref="StreamInfo"/> instance with information about the stream.</returns>
+        /// <remarks>https://redis.io/topics/streams-intro</remarks>
+        StreamInfo StreamInfoGet(RedisKey key, CommandFlags flags = CommandFlags.None);
```
IMO, this should be just `StreamInfo` (we don't have `Get` as a convention or anything - just in the few places it makes sense).

A few options to consider:
1. `StreamGetInfo`, `StreamGetGroupInfo`, `StreamGetConsumerInfo`
2. `StreamInfo`, `StreamGroupInfo`, `StreamConsumerInfo`
3. `StreamGetInfo`, `StreamGetGroups`, `StreamGetConsumers`

They're all `XINFO` under the covers, and IMO it'd be nice to indicate that in naming if we can - I'm personally a fan of option 2. Thoughts?

ttingen replied:
I like #2 as well.

NickCraver commented on StackExchange.Redis/StackExchange/Redis/Interfaces/IDatabase.cs:
```diff
@@ -1389,6 +1389,250 @@ public interface IDatabase : IRedis, IDatabaseAsync
+        /// <returns>An instance of <see cref="StreamPendingInfo"/>. <see cref="StreamPendingInfo"/> contains the number of pending messages, the highest and lowest ID of the pending messages, and the consumers with their pending message count.</returns>
+        /// <remarks>The equivalent of calling XPENDING key group.</remarks>
+        /// <remarks>https://redis.io/commands/xpending</remarks>
+        StreamPendingInfo StreamPendingInfoGet(RedisKey key, RedisValue groupName, CommandFlags flags = CommandFlags.None);
```
`StreamPending`?

ttingen replied:
+1

NickCraver commented on StackExchange.Redis/StackExchange/Redis/Interfaces/IDatabase.cs:
```diff
@@ -1389,6 +1389,250 @@ public interface IDatabase : IRedis, IDatabaseAsync
+        /// <returns>An instance of <see cref="StreamPendingMessageInfo"/> for each pending message.</returns>
+        /// <remarks>Equivalent of calling XPENDING key group start-id end-id count consumer-name.</remarks>
+        /// <remarks>https://redis.io/commands/xpending</remarks>
+        StreamPendingMessageInfo[] StreamPendingMessageInfoGet(RedisKey key, RedisValue groupName, RedisValue minId, RedisValue maxId, int count, RedisValue consumerName, CommandFlags flags = CommandFlags.None);
```
`StreamPendingMessages`?

ttingen replied:
+1

NickCraver commented on StackExchange.Redis/StackExchange/Redis/Interfaces/IDatabase.cs:
```diff
@@ -1389,6 +1389,250 @@ public interface IDatabase : IRedis, IDatabaseAsync
+        /// <returns>An instance of <see cref="RedisStream"/> for each stream.</returns>
+        /// <remarks>Equivalent of calling XREAD COUNT num STREAMS key1 key2 id1 id2.</remarks>
+        /// <remarks>https://redis.io/commands/xread</remarks>
+        RedisStream[] StreamRead(IList<KeyValuePair<RedisKey, RedisValue>> streamAndIdPairs, int? countPerStream = null, CommandFlags flags = CommandFlags.None);
```
@mgravell thoughts here? We don't have this type of input type (AFAIK) anywhere else in the API, the closest equivalent is `SortedSetEntry[]` for adds, which incurs the array overhead you'd likely want to avoid in streams in general...suggestions?

ttingen replied:
I’ve always thought this one looked clunky. An input type tailored for this method would look a lot cleaner.

Something like this?
StreamRead(StreamIdPair[] streamIdPairs, int? countPerStream, ...)

magmasystems commented:
Thanks for this. I did some initial work in my own framework (sitting on top of StackExchange.Redis) to put in support for Streams, and all I had to change in the Redis 1.26 source was to add the Stream enum to RedisType.cs. Everything else was handled by my own framework, and a healthy dose of calling Execute() and ExecuteAsync() and hand-coding the stream parsing. I am looking forward to the official support for streams that you guys are implementing.

ttingen commented:
@magmasystems, thanks for mentioning RedisType. That needed to be added.

magmasystems commented:
No probs, Todd. I had to do it because I wrote a custom WPF-based GUI and Test Harness for Redis that needed to query the data type of what a key pointed to in order to have a custom WPF data template.

mgravell commented:
Presumably the "mutate" commands are "master only". It may be worthwhile doing a quick review and adding those ones to `IsMasterOnly` in `Message.cs` - that will avoid them being sent incorrectly to replica (slave) nodes.

NickCraver commented:
@mgravell It shouldn't stop this PR because it's really a more global/separate issue, but note #354 is related here - I added some notes/a branch there.

ttingen commented:
@NickCraver, 4 more overloads removed. The streamEntryId parameter was renamed to messageId for the StreamAdd overloads. 

Also added a StreamIdPair struct and updated one of the StreamRead overloads to use that instead of IList<KeyValuePair<RedisKey, RedisValue>>. Looks a lot cleaner now.

magmasystems commented:
@ttingen - Just wondering if you have ReSharper and are doing any C# 7 refactoring as you go along, or eliminating any of the ReSharper "yellows"?

NickCraver commented on StackExchange.Redis/StackExchange/Redis/Interfaces/IDatabase.cs:
```diff
@@ -1389,6 +1389,225 @@ public interface IDatabase : IRedis, IDatabaseAsync
+        /// <param name="flags">The flags to use for this operation.</param>
+        /// <returns>The message IDs for the messages successfully claimed by the given consumer.</returns>
+        /// <remarks>https://redis.io/topics/streams-intro</remarks>
+        RedisValue[] StreamClaimMessagesReturningIds(RedisKey key, RedisValue consumerGroup, RedisValue claimingConsumer, long minIdleTimeInMs, RedisValue[] messageIds, CommandFlags flags = CommandFlags.None);
```
I was thinking about these, and it finally occurs to me: everything is `Messages`. We need not say it. For example, we don't call it `StreamAddMessage()`, `StreamAcknowledgeMessages()`, etc.

What if we shortened these `StreamClaim()` and `StreamClaimIdsOnly()`? Thoughts?

ttingen replied:
I like it, making those changes now.

NickCraver commented on StackExchange.Redis/StackExchange/Redis/RedisDatabase.cs:
```diff
@@ -1579,6 +1579,330 @@ public IEnumerable<SortedSetEntry> SortedSetScan(RedisKey key, RedisValue patter
+                RedisCommand.XGROUP,
+                key,
+                groupName,
+                readFrom);
```
Shouldn't this match non-async?

ttingen replied:
Indeed it should. That's what I get for not testing the Async versions yet...

NickCraver commented on StackExchange.Redis/StackExchange/Redis/Interfaces/IDatabase.cs:
```diff
@@ -1389,6 +1389,225 @@ public interface IDatabase : IRedis, IDatabaseAsync
+        /// <param name="flags">The flags to use for this operation.</param>
+        /// <returns>Returns the number of messages successfully deleted from the stream.</returns>
+        /// <remarks>https://redis.io/topics/streams-intro</remarks>
+        long StreamMessagesDelete(RedisKey key, RedisValue[] messageIds, CommandFlags flags = CommandFlags.None);
```
I think we can call this `StreamDelete`, as the `messageIds` are required and make it clear it's not deleting a stream.

ttingen commented:
@NickCraver, thanks and it's been a pleasure.

mgravell commented:
Heya; there's a *lot* of work here, so once again: thanks hugely, this PR is great. I've been deliberately and consciously avoiding jumping in so that we had a second set of "clean eyes" for a final review.

Firstly: the code is looking in great shape, thanks. It is also without doubt the single biggest PR we've received for SE.Redis, so: (awards sticky gold star)

There's a couple of things I want to quickly ask about before we click that final merge button, if you don't mind (but: this is definitely very close!)

Firstly, `XREAD`; my understanding is that there is a multi-key version of this that allows the consumer to fetch from multiple streams; I couldn't see an API that surfaces this - is that intentional?

---

Secondly, I'm still a little ... "uncomfortable" about the ambiguity of message id; in some places (mostly: add) they are a string, and in other (most) places they are `RedisValue`. I have seen your comment above about wanting to default to `"*"`, but ... I think this is actually an implementation detail that we don't need to surface in the API. I suspect the most appropriate API here is probably:

```c#
RedisValue messageId = default
```

if that compiles, or

```c#
RedisValue? messageId = null
```

if it doesn't; then just check in the *implementation* code whether the `messageId` is nil; if it is, then the *implementation* code can substitute the wildcard.

---

This made me think about `StreamConstants` in general; it looks like almost all of these are internal implementation details, which is good. I'm a little suspicious of `StreamConstants.ReadMinValue` etc for similar reasons to `messageId` above - I wonder if that should *also* just default to a nil `RedisValue` using the same approach used by `messageId` above. Essentially: "if you don't supply this, it'll use the min/max", which is *I think* the intent? If not, please let me know!

With that done, I wonder whether `StreamConstants` can be made `internal` - as far as I can see, there's nothing left that isn't already handled by API options rather than magic values on the API.

Thoughts?

magmasystems commented:
Apologies for butting in here with something that may be slightly off-topic to the PR.

I have spent a good part of my working life architecting and developing low-latency trading systems and other systems that use messaging. Usually that messaging is done through a JMS or AMQP interface over a bus like Tibco EMS, RabbitMQ, or ActiveMQ.

I think that Marc and Nick's philosophy for StackExchange.Redis is to maintain a fairly thin wrapper over the Redis API, and not to introduce any opinionated higher-level framework over StackExchange.Redis. But I am wondering if there may be an appetite for some higher-level framework that sits over the streams. For example, a RedisStreamProducer that can publish a stream to a "topic" (where the "topic" is the key of a stream) and a RedisStreamConsumer which fires a C# event when data is read from a "topic".

When I first tested out the new Redis streams, I wrote something like that in my own code, and I did not consider altering the StackExchange.Redis code in any way, given the fact that I view StackExchange.Redis as a wrapper. If Marc/Nick thinks that a higher-level framework is interesting, I can try to put something together. Otherwise, I will just keep the JMS stuff in my own code,

mgravell commented:
@magmasystems  my main concern there is: is there an existing that would allow it to be consumed from multiple places? or would it have to be provider specific. If it is provider specific, I'd rather keep it out of the main lib purely for dependency reasons; there's scope for sibling-libraries, though, that re-expose the "raw" library in the context of what that means for "tool X"

NickCraver commented:
> I think this is actually an implementation detail that we don't need to surface in the API. I suspect the most appropriate API here is probably

Agreed

> I wonder whether StreamConstants can be made `internal`

Also agree, same with making them null by default and min/max in the implementation.

@magmasystems I think this would be where another library sits atop the APIs here, could be my view there. We could always add more concepts later if doing so was very appealing.

ttingen commented:
> Firstly, XREAD; my understanding is that there is a multi-key version of this that allows the consumer to fetch from multiple streams; I couldn't see an API that surfaces this...

The StreamRead overload that takes an array of StreamIdPair values is the API that reads from multiple streams. It's the one that previously used an IList<KeyValuePair<RedisKey, RedisValue>> as an input parameter.

`RedisStream[] StreamRead(StreamIdPair[] streamIdPairs, int? countPerStream = null, ...)`

The current version looks a lot cleaner than the original but perhaps the naming of the StreamIdPair struct could be improved. I started with `StreamAndIdPair` but didn't like that name too much. Thoughts?

The StreamReadGroup API does however only expose the ability to read from a single stream at a time. Do you want me to add an overload that allows reading multiple streams there as well?

> ...about the ambiguity of message id...

Totally agree, I will make the change to `RedisValue? messageId = null`.  The `RedisValue messageId = default` wouldn't compile with the current language setting (7.0) for the library.

> This made me think about StreamConstants [being changed to] `internal`...

The only other two "constants" I was concerned about were "$" (new messages) and ">" (undelivered messages), they would be used with the `XGROUP CREATE` and `XREADGROUP` commands. But those values would work as sensible defaults for those APIs as well. 

I will update the documentation to include details on the default values.

@mgravell/@NickCraver, let me know what you think about `StreamIdPair` and the additional `StreamReadGroup` overload. I will start making these changes later this evening.

ttingen commented:
Additionally, I only created the single-stream-reading version of StreamReadGroup to keep the API simple and straightforward. 

I felt it would minimize confusion to only expose the single stream version. For example, to read multiple streams at once using `XREADGROUP`, each stream would need to have created a Consumer Group with the same name. Not that setting up the consumer groups is difficult, I just like the "don't make me think" mentality and decided to keep it simple.

But I'll certainly add the overload if you want it there.

mgravell commented:
>  I started with StreamAndIdPair but didn't like that name too much. Thoughts?

Naming is hard; it does what it says, I guess; I think it is fine - thanks for the clarification

> EDIT: I gave this more consideration and I will add the overload. Since this is a more generalized library, I shouldn’t limit the scope to just my use cases or sensibilities.

Fair enough; I wasn't going to ask you to add it, but if it is there : great

Will glance over code again :)

mgravell commented:
In a bunch of the tests, things there's a lot of changes like:

```
- db.StreamCreateConsumerGroup(key, groupName, StreamConstants.ReadMinValue);
+ db.StreamCreateConsumerGroup(key, groupName, "-");
```

Just so I understand: is this because you're being explicit? i.e. would this be semantically equivalent? or is it different?

```
db.StreamCreateConsumerGroup(key, groupName);
```

(just so I understand)

mgravell commented:
API looks a lot cleaner, thanks!

ttingen commented:
I defaulted the readFrom parameter to `"$"` (only read new messages) for  `StreamCreateConsumerGroup` and `">"` (only read undelivered messages) for `StreamReadGroup`. 

Those are the options used in the first mentions of the `XGROUP CREATE` and `XREADGROUP` commands in the [intro to streams](https://redis.io/topics/streams-intro) docs and IIRC those options are only valid for those commands. I believe they are reasonable defaults but they can be overridden by passing "-" (start from the lowest ID in the stream) or by passing an actual ID from the stream. Many of the tests for Consumer Groups needed to read from the beginning of the stream so I was passing "-" when creating the consumer group. 

Another interesting caveat when using `"-"` with `XREADGROUP` ... when used in conjunction with a consumer group that was created to only read new messages, passing `"-"` to `XREADGROUP` will read from the beginning of the stream of messages created AFTER the consumer group was created, not the actual beginning of the stream.

Consumer Groups is definitely one of the more complex features of Redis Streams. I suppose the API could be modeled to expose each of the available options as a different method, at least that would be very explicit. Leaving it as-is certainly enables a lot of flexibility in how each method can be utilized but proper use requires a sound understanding of the underlying commands and their options.

Or maybe, instead of explicit methods, the API takes and options struct that would detail how the command should be constructed. Hmm, I'll put some thought into how this functionality could be made more intuitive in the API.

I updated the method documentation to highlight the default value that would be passed to the command if a value is not supplied for the parameter.

...Sorry for lengthy response...I will work on the remaining overload tomorrow morning.

mgravell commented:
It is also entirely possible that what you have is perfectly fine and usable. Trying to make an API completely idiot-proof can also mean bending it out of shape; I'm not sure that is necessarily an improvement :) I'd welcome any more thoughts you have on this, but "I've considered it, and I'm happy with it like this" might also be fine and reasonable.

ttingen commented:
> Trying to make an API completely idiot-proof can also mean bending it out of shape.

Great point. :)

ttingen commented:
@NickCraver I just [added a commit](https://github.com/ttingen/StackExchange.Redis/commit/7d283c1862fdc3b7692a772633d98c7413762774) to my fork that adds the multi-stream reading overload for StreamReadGroup. I also updated some documentation and parameter names to be "more correct". Check it out and let me know if you want me to do another PR to add those changes.

mgravell commented:
merged into 2.0 - the only awkward bit  was the temp/fake multi-bulk around `ParseRedisStreamEntries` - I *think* I've converted that code correctly. Thanks again.

(/cc @ttingen  @NickCraver )

ttingen commented:
@mgravell  @NickCraver I tinkered with passing an options struct to the StreamCreateConsumerGroup method, I think I like it. Now there's no need for the user to know any of the magic strings. What do you think?

`db.StreamCreateConsumerGroupAsync("events", "consumerGroup", GroupCreateOptions.BeginningOfStream);`

![image](https://user-images.githubusercontent.com/1170315/42291665-38f56b0c-7f9b-11e8-8d25-c8a7028a3f3a.png)

![image](https://user-images.githubusercontent.com/1170315/42291753-b8f209b4-7f9b-11e8-9289-9f25c6359fbc.png)

mgravell commented:
I quite like that, yes. Seems clean. I'm in favour of a new PR to apply
that. I can't speak for Nick, of course.

On Wed, 4 Jul 2018, 20:11 Todd Tingen, <notifications@github.com> wrote:

> @mgravell <https://github.com/mgravell> @NickCraver
> <https://github.com/NickCraver> I tinkered with passing an options struct
> to the StreamCreateConsumerGroup method, I think I like it. Now there's no
> need for the user to know any of the magic strings. What do you think?
>
> db.StreamCreateConsumerGroupAsync("events", "consumerGroup",
> GroupCreateOptions.BeginningOfStream);
>
> [image: image]
> <https://user-images.githubusercontent.com/1170315/42291665-38f56b0c-7f9b-11e8-8d25-c8a7028a3f3a.png>
>
> [image: image]
> <https://user-images.githubusercontent.com/1170315/42291753-b8f209b4-7f9b-11e8-9289-9f25c6359fbc.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/StackExchange/StackExchange.Redis/pull/860#issuecomment-402545127>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABDsNbeioh5-oDgp0TTJHDrxo5mrpYJks5uDRNygaJpZM4UqtE5>
> .
>

ttingen commented:
@NickCraver @mgravell And thanks again for taking so much time to work with me on this PR, I know you two are very busy individuals.

mgravell commented:
Meh, everyone is busy :) But *because* we've been busy doing 47 other things is why we live so much that someone else found the time to look at this and do the thinking for us :)

mgravell commented:
blocking commands are incompatible with the connection-sharing approach used by SE.Redis; my instinct would be "no"; see the discussion around "blocking pops" [here](https://stackexchange.github.io/StackExchange.Redis/PipelinesMultiplexers#multiplexing) for the same problem.
